activitysim/abm/models/tour_mode_choice_reallocation.py

```diff
@@ -0,0 +1,653 @@
+    """
+    Adds and codes household auto sufficiency based on household auto ownership and number of drivers in the household
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+
+    Returns
+    -------
+    households : pd.Dataframe
+        Household table with auto sufficiency column added
+
+    """
+    households['auto_sufficiency'] = 'suff'
+    households.loc[
+        households.auto_ownership.between(0, households.num_drivers, 'neither'), 'auto_sufficiency'] = 'insuff'
+    households.loc[households.auto_ownership == 0, 'auto_sufficiency'] = 'nocar'
+
+    return households
+
+
+def filter_for_multi_drive_tour_households(households, persons, tours):
+    """
+    Filters tours down to only those from auto insufficient households with multiple drivers who made SOV-based tours
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+    persons : pd.Dataframe
+    tours : pd.Dataframe
+
+    Returns
+    -------
+    tours_insuff_mult_drv : pd.Dataframe
+        Tours from auto insufficient households with multiple drivers who made SOV-based tours
+
+    """
+    households = code_household_auto_sufficiency(households)
+
+    # Isolate for SOV-based tours by driving aged persons in auto insufficient households.
+    households_insuff = households[households.auto_sufficiency == 'insuff']
+    person_drivers = persons[persons.is_driving_age]
+    # TODO: Expand criteria to cover joint tours once a reallocation method can be developed for it
+    tours_insuff_drv = tours[
+        tours.household_id.isin(households_insuff.index) &
+        tours.person_id.isin(person_drivers.index) &
+        (tours.tour_mode == 'DRIVEALONEFREE')]
+
+    # Get per household count of driver who made SOV-based tours in auto insufficient households
+    unique_pers_insuff_drv = tours_insuff_drv[['household_id', 'person_id']].drop_duplicates()
+
+    count_pers_insuff_drv = unique_pers_insuff_drv.groupby('household_id').person_id.count()
+    count_pers_insuff_drv = count_pers_insuff_drv[count_pers_insuff_drv > 1]
+
+    # Isolate for tours from auto insufficient households with multiple drivers who made SOV-based tours
+    tours_insuff_mult_drv = tours_insuff_drv[tours_insuff_drv.household_id.isin(count_pers_insuff_drv.index)]
+    tours_insuff_mult_drv = tours_insuff_mult_drv.sort_values(['household_id', 'start', 'end'])
+
+    return tours_insuff_mult_drv
+
+
+def create_time_grid(filtered_tours):
+    """
+    Creates per tour time grid flagging the 30-min intervals each tour occupies
+
+    Parameters
+    ----------
+    filtered_tours : pd.Dataframe
+        Tours from auto insufficient households with multiple drivers who made SOV-based tours
+
+    Returns
+    -------
+    tour_time_grid : pd.Dataframe
+        Per tour table with a grid of 48 30-min intervals flagged based on tour start and end time
+
+    """
+    tour_time_grid = pd.DataFrame({'household_id': filtered_tours['household_id']})
+    for i in range(1, 49):
```

Comment:
@TianYangLin Are the time intervals always 30-min intervals? If the duration of each time interval is configurable in ActivitySim, then we cannot hard code the number of intervals in this module.

Reply:
The 30-min intervals are not hard coded into the model. The time interval resolution of the model is read in implicitly through the tour departure and duration alternatives (tdd_alt). Will change this section to use the tdd_alt.